### PR TITLE
New backoffice: examine management controller

### PR DESCRIPTION
**Pull request by Zeegaan**
# Notes
Beware that the serialization of Dictionary<string, object> does not work for System.Text.Json in .NET 6, this has however been fixed in .NET 7, see more on this issue: https://github.com/dotnet/runtime/issues/67588
## What does this mean for us though?
Well this means that most of our actions just straight-up throw exceptions for now. 
When testing, the Dictionary<string, object> should be commented out from the ViewModel, so we can assert at least the rest of the logic works as intended 👍 
## Other changes
- Added `ExamineManagerService` to handle logic for validating indexes, searchers & populators instead of that logic being in the controller. Could argue for a better service name, but can't think of a better name 😅 
- Created the `ExamineIndexViewModel`
- Created an `ExamineIndexViewModelFactory` to handle logic of creating the view model.

# How to test
- Comment out the `public IReadOnlyDictionary<string, object?> ProviderProperties { get; set; } = null!;` from ExamineIndexViewModel.cs
- Try and build the project, you will get another error in the ExamineIndexViewModel factory as we now are trying to map something that no longer exists, comment out line 42 aswell and you're good to go!
- Test the different endpoints using this postman collection and assert they do as described: https://www.dropbox.com/s/76zkugv38f2lc8x/ExamineManagement.postman_collection.json?dl=0

### CODE REVIEW

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    ///     have a listener for the index op complete so we'll just check if that key is no longer there in the runtime cache
+    /// </remarks>
+    [HttpGet("hasIndexRebuilt")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ExamineIndexViewModel), StatusCodes.Status200OK)]
+    // This endpoint for now will throw errors if the ExamineIndexViewModel ever has providerProperties defined
+    // This is because System.Text.Json cannot serialize dictionary<string, object>
+    // This has been fixed in .NET 7, so this will work when we upgrade: https://github.com/dotnet/runtime/issues/67588
+    public async Task<ActionResult<ExamineIndexViewModel?>> HasIndexRebuilt(string indexName)
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs:45]: What is the purpose of this endpoint? It feels kind of weirdly specific to a very specific use case, but maybe it needs to be broader.

In principle the `Indexes` endpoint could be sufficient for querying, since we're not likely to have a whole lot of indexes per site and thus the overhead of returning all indexes is negligible. However, it wouldn't hurt to have an endpoint for retrieving a single index - maybe named `Index` . 

I'd also argue that the "is rebuilding" magic with the runtime cache belongs in the view model factory rather than the controller, and that it is relevant info for any index view model, whether one is querying all indexes or a single index.

So to sum up, here is something to consider:

1. Move the "is rebuilding" detection to the view model factory.
2. Rename this endpoint to `Index` and let it work like the `Indexes` endpoint, just for a specific index name.
3. Leave the populator validation to the `RebuildIndex` endpoint.

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildIndexExamineManagementController.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        // remove it in case there's a handler there already
+        index!.IndexOperationComplete -= Indexer_IndexOperationComplete;
+
+        //now add a single handler
+        index.IndexOperationComplete += Indexer_IndexOperationComplete;
+
+        try
+        {
+            var cacheKey = "temp_indexing_op_" + index.Name;
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildIndexExamineManagementController.cs:78]: It would be nice with a small service that could handle setting, clearing and detecting `temp_indexing_op_` from the runtime cache ... it could then eventually be turned into something less rudimentary without breaking the examine management implementation.

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildIndexExamineManagementController.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        }
+        catch (Exception ex)
+        {
+            //ensure it's not listening
+            index.IndexOperationComplete -= Indexer_IndexOperationComplete;
+            _logger.LogError(ex, "An error occurred rebuilding index");
+            var invalidModelProblem = new ProblemDetails
+            {
+                Title = "Index cannot be rebuilt",
+                Detail = $"The index {index.Name} cannot be rebuilt because it does not have an associated {typeof(IIndexPopulator)}",
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildIndexExamineManagementController.cs:94]: This can't possibly be the correct explanation to put in the `Detail` property - if we get here, there most certainly is an associated populator for the index (see guarding code above)

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    [HttpGet("searchers")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(PagedViewModel<ExamineIndexModel>), StatusCodes.Status200OK)]
+    // This endpoint for now will throw errors if the ExamineIndexViewModel ever has providerProperties defined
+    // This is because System.Text.Json cannot serialize dictionary<string, object>
+    // This has been fixed in .NET 7, so this will work when we upgrade: https://github.com/dotnet/runtime/issues/67588
+    public async Task<PagedViewModel<ExamineIndexViewModel>> GetSearcherDetails(int skip, int take)
+    {
+        var searchers = new List<ExamineIndexViewModel>(
+            _examineManager.RegisteredSearchers.Select(searcher => new ExamineIndexViewModel { Name = searcher.Name })
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs:31]: Should return a dedicated view model instead of reusing `ExamineIndexViewModel` here

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    ///     Get the details for searchers
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("searchers")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(PagedViewModel<ExamineIndexModel>), StatusCodes.Status200OK)]
+    // This endpoint for now will throw errors if the ExamineIndexViewModel ever has providerProperties defined
+    // This is because System.Text.Json cannot serialize dictionary<string, object>
+    // This has been fixed in .NET 7, so this will work when we upgrade: https://github.com/dotnet/runtime/issues/67588
+    public async Task<PagedViewModel<ExamineIndexViewModel>> GetSearcherDetails(int skip, int take)
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs:28]: `GetSearcherDetails` should probably be `Searchers` to follow the modus of naming endpoints and routes the same?

File: src/Umbraco.Cms.ManagementApi/Factories/ExamineIndexViewModelFactory.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    public ExamineIndexViewModel Create(IIndex index)
+    {
+        var indexName = index.Name;
+
+        IIndexDiagnostics indexDiag = _indexDiagnosticsFactory.Create(index);
+
+        Attempt<string?> isHealthy = indexDiag.IsHealthy();
+
+        var properties = new Dictionary<string, object?>
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Factories/ExamineIndexViewModelFactory.cs:27]: Is there a specific reason why we need these two properties in a generic dictionary instead of declaring them as explicit properties on `ExamineIndexViewModel`?

File: src/Umbraco.Cms.ManagementApi/Factories/ExamineIndexViewModelFactory.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        };
+
+        foreach (KeyValuePair<string, object?> p in indexDiag.Metadata)
+        {
+            properties[p.Key] = p.Value;
+        }
+
+        var indexerModel = new ExamineIndexViewModel
+        {
+            Name = indexName,
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Factories/ExamineIndexViewModelFactory.cs:40]: Maybe extract and return the associated searcher name here? It would be nice to have it be an explicit part of the output, so you know what searcher name to use when searching the index ; right now you have to deduct that you need to use the index name as searcher name.

File: src/Umbraco.Cms.ManagementApi/Services/ExamineManagerService.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+﻿using Examine;
+using Umbraco.Cms.Infrastructure.Examine;
+
+namespace Umbraco.Cms.ManagementApi.Services;
+
+public class ExamineManagerService : IExamineManagerService
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Services/ExamineManagerService.cs:6]: I think the only justification for this class is the `TryFindSearcher` method. The rest of the methods are just renaming proxied methods on `IExamineManager` or `IIndexRebuilder`.

I would consider reshaping this class to only be used for finding a searcher, and leave the rest of the implementations to be used as-is from their respective services.

I'll add a few more comments to this file, they can be disregarded if the class is reshaped 😄

File: src/Umbraco.Cms.ManagementApi/Services/ExamineManagerService.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+            return _examineManager.TryGetSearcher(searcherName, out searcher);
+        }
+
+        searcher = index.Searcher;
+        return true;
+    }
+
+    public bool ValidateIndex(string indexName, out IIndex? index)
+    {
+        index = null;
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Services/ExamineManagerService.cs:32]: No need to assign to `index` here, `examineManager.TryGetIndex` handles it.

File: src/Umbraco.Cms.ManagementApi/Services/ExamineManagerService.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+        return _examineManager.TryGetIndex(indexName, out index);
+    }
+
+    public bool ValidatePopulator(IIndex index) => _indexRebuilder.CanRebuild(index.Name);
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Services/ExamineManagerService.cs:37]: I'm not a fan of the method name here - it assumes too much. It is dangerous to assume that `_indexRebuilder.CanRebuild` can only return false if there is no populator assigned to the index. At the moment the assumption is true, but down the road no-one knows what might cause `_indexRebuilder.CanRebuild` to return false.

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildIndexExamineManagementController.cs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    /// <summary>
+    ///     Rebuilds the index
+    /// </summary>
+    /// <param name="indexName"></param>
+    /// <returns></returns>
+    [HttpPost("rebuildIndex")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(OkResult), StatusCodes.Status200OK)]
+    public async Task<IActionResult> RebuildIndex(string indexName)
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildIndexExamineManagementController.cs:40]: I wonder if this endpoint should simply be `Rebuild`? The other endpoints have no "Index" postfix.

File: src/Umbraco.Cms.ManagementApi/ViewModels/ExamineManagement/SearchResultViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    public string? Id { get; set; }
+
+    [JsonPropertyName("score")]
+    public float Score { get; set; }
+
+    [JsonPropertyName("fieldCount")]
+    public int FieldCount => Values?.Count ?? 0;
+
+    [JsonPropertyName("values")]
+    public IReadOnlyDictionary<string, IReadOnlyList<string>>? Values { get; set; }
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/ViewModels/ExamineManagement/SearchResultViewModel.cs:17]: This should be remodelled as strongly typed models and maybe renamed to `Fields`. Something like:

```csharp
[JsonPropertyName("fields")]
public IEnumerable<Field> Fields { get; set; }

...

public class Field 
{
  public string Name { get; }
  
  public IEnumerable<string> Values { get; }
}

```

**Review by kjac — CHANGES_REQUESTED** [review]
@Zeegaan I have left a few review comments for you to consider.

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs (hunk 1 above)
Comment (Zeegaan) [reply src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs]: Great input, I've tried moving this logic to the ViewModelFactory, but a bit unsure of how to handle it properly, if the index is actually rebuilding, for now, we're throwing a OperationCancelledException

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
+﻿using Examine;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Umbraco.Cms.ManagementApi.Factories;
+using Umbraco.Cms.ManagementApi.ViewModels.ExamineManagement;
+
+namespace Umbraco.Cms.ManagementApi.Controllers.ExamineManagement;
+
+[ApiVersion("1.0")]
+public class HasIndexRebuiltExamineManagementController : ExamineManagementControllerBase
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs:10]:
```suggestion
public class IndexExamineManagementController : ExamineManagementControllerBase
```

Rename controller to match new operation name

Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs:10]: This controller should be renamed too, to match the new operation name - rename to `IndexExamineManagementController`

File: src/Umbraco.Cms.ManagementApi/Factories/ExamineIndexViewModelFactory.cs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+        _indexDiagnosticsFactory = indexDiagnosticsFactory;
+        _indexRebuilder = indexRebuilder;
+        _temporaryIndexingService = temporaryIndexingService;
+    }
+
+    public ExamineIndexViewModel Create(IIndex index)
+    {
+        if (_temporaryIndexingService.Detect(index.Name))
+        {
+            throw new OperationCanceledException("The index is still rebuilding, so could not get it");
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Factories/ExamineIndexViewModelFactory.cs:26]: Instead of throwing an exception here, maybe return an `ExamineIndexViewModel` that has `"Rebuilding"` as  `HealthStatus` and `0` as `DocumentCount` and `FieldCount` ?

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs (hunk 1 above)
Comment (kjac) [reply src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs]: 👍 

I left a suggestion [here](https://github.com/umbraco/Umbraco-CMS/pull/12981/files#r979194417) 😄

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/IndexesExamineManagementController.cs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        _examineIndexViewModelFactory = examineIndexViewModelFactory;
+    }
+
+    /// <summary>
+    ///     Get the details for indexers
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("indexes")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(PagedViewModel<ExamineIndexModel>), StatusCodes.Status200OK)]
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/IndexesExamineManagementController.cs:32]:
```suggestion
    [ProducesResponseType(typeof(PagedViewModel<ExamineIndexViewModel>), StatusCodes.Status200OK)]
```

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+    public SearchersExamineManagementController(IExamineManager examineManager) => _examineManager = examineManager;
+
+    /// <summary>
+    ///     Get the details for searchers
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("searchers")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(PagedViewModel<ExamineIndexModel>), StatusCodes.Status200OK)]
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs:24]:
```suggestion
    [ProducesResponseType(typeof(PagedViewModel<SearcherViewModel>), StatusCodes.Status200OK)]
```

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildExamineManagementController.cs [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+            //ensure it's not listening
+            index.IndexOperationComplete -= Indexer_IndexOperationComplete;
+            _logger.LogError(ex, "An error occurred rebuilding index");
+            var invalidModelProblem = new ProblemDetails
+            {
+                Title = "Index cannot be rebuilt",
+                Detail = $"The index {index.Name} cannot be rebuilt because it does not have an associated {typeof(IIndexPopulator)}",
+                Status = StatusCodes.Status400BadRequest,
+                Type = "Error",
+            };
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildExamineManagementController.cs:93]: At this point the index *will* have an index populator registered, otherwise this action would have failed earlier. I suggest referring to the log for details.

```suggestion
            var invalidModelProblem = new ProblemDetails
            {
                Title = "Index could not be rebuilt",
                Detail = $"The index {index.Name} could not be rebuild. Check the log for details on this error.",
                Status = StatusCodes.Status400BadRequest,
                Type = "Error",
            };
```

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildExamineManagementController.cs [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+            return await Task.FromResult(BadRequest(invalidModelProblem));
+        }
+
+        if (!_indexRebuilder.CanRebuild(index.Name))
+        {
+            var invalidModelProblem = new ProblemDetails
+            {
+                Title = "Could not validate the populator",
+                Detail = $"The index {index?.Name} cannot be rebuilt because we could not validate its associated {typeof(IIndexPopulator)}",
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/RebuildExamineManagementController.cs:59]: Nitpicking here, but I would use past tense consequently for the message.

```suggestion
                Detail = $"The index {index?.Name} could not be rebuilt because we could not validate its associated {typeof(IIndexPopulator)}",
```

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchExamineManagementController.cs [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+                .CreateQuery()
+                .NativeQuery(query)
+                .Execute(QueryOptions.SkipTake(skip, take));
+        }
+        catch (ParseException)
+        {
+            var invalidModelProblem = new ProblemDetails
+            {
+                Title = "Could not parse the query",
+                Detail = "Parser cannot parse the query, this can happen as en example if your query starts with a *)",
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchExamineManagementController.cs:61]: Nitpicking about tense and wording again. Sorry 😅 

```suggestion
                Detail = "Parser could not parse the query. Please double check if the query is valid. Sometimes this can also happen if your query starts with a wildcard (*)",
```

File: src/Umbraco.Cms.ManagementApi/ManagementApiComposer.cs [hunk 20]
```diff
@@ -35,7 +35,9 @@ public void Compose(IUmbracoBuilder builder)
 
         builder
             .AddNewInstaller()
-            .AddUpgrader();
+            .AddUpgrader()
+            .AddServices()
+            .AddFactories();
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/ManagementApiComposer.cs:40]: Single entry points for factories and services will eventually grow rather large and it'll be hard to figure out which factories and services are used for what. 

I believe `AddServices()` and `AddFactories()` should be replaced by one or more Examine dedicated `IUmbracoBuilder` extensions - i.e. `AddExamineManagement()`  (like `AddNewInstaller()` and `AddUpgrader()`).

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    ///     This is kind of rudimentary since there's no way we can know that the index has rebuilt, we
+    ///     have a listener for the index op complete so we'll just check if that key is no longer there in the runtime cache
+    /// </remarks>
+    [HttpGet("index")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ExamineIndexViewModel), StatusCodes.Status200OK)]
+    // This endpoint for now will throw errors if the ExamineIndexViewModel ever has providerProperties defined
+    // This is because System.Text.Json cannot serialize dictionary<string, object>
+    // This has been fixed in .NET 7, so this will work when we upgrade: https://github.com/dotnet/runtime/issues/67588
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/HasIndexRebuiltExamineManagementController.cs:38]: This comment no longer applies after removing the `ProviderProperties` from `ExamineIndexViewModel`
```suggestion
```

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/IndexesExamineManagementController.cs [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+    // This endpoint for now will throw errors if the ExamineIndexViewModel ever has providerProperties defined
+    // This is because System.Text.Json cannot serialize dictionary<string, object>
+    // This has been fixed in .NET 7, so this will work when we upgrade: https://github.com/dotnet/runtime/issues/67588
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/IndexesExamineManagementController.cs:35]: This comment no longer applies after removing the `ProviderProperties` from `ExamineIndexViewModel`
```suggestion
```

File: src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    // This endpoint for now will throw errors if the ExamineIndexViewModel ever has providerProperties defined
+    // This is because System.Text.Json cannot serialize dictionary<string, object>
+    // This has been fixed in .NET 7, so this will work when we upgrade: https://github.com/dotnet/runtime/issues/67588
```
Comment (kjac) [inline src/Umbraco.Cms.ManagementApi/Controllers/ExamineManagement/SearchersExamineManagementController.cs:27]: This comment no longer applies after removing the `ProviderProperties` from `ExamineIndexViewModel`
```suggestion
```

**Review by kjac — CHANGES_REQUESTED** [review]
@Zeegaan great work 😄 I've added some more comments and considerations.

**Review by kjac — APPROVED** [review]
💪
